gcc/d/d-decls.cc

```diff
@@ -101,7 +101,38 @@ VarDeclaration::toSymbol()
 	  return csym;
 	}
 
-      csym = new Symbol();
+      // Use only same Symbol for VarDeclarations with same mangle
+      if (!this->mangleOverride && isDataseg () && !(storage_class & STCextern))
+	{
+	  static hash_map<nofree_string_hash, Symbol *> duplicate_map(251);
+	  const char *vname = mangle(this);
+	  bool existed;
+
+	  Symbol **entry = &duplicate_map.get_or_insert(vname, &existed);
+	  if (!existed)
+	    {
+	      csym = new Symbol();
+	      *entry = csym;
+	    }
+	  else
+	    {
+	      csym = *entry;
+	      if (!csym->Sloc.equals(loc))
+		{
+		  error ("Found different variables with identical mangle: %s"
+		    " at %s and %s", vname, loc.toChars(), csym->Sloc.toChars());
```

Comment:
> I wonder whether we could acheive the same by hooking up lang_identifier with binding_level where we keep track of all active binding levels DECL_NAME is being used in. And throw an error if it shows twice in the same scope.

Maybe I misunderstood, but I don't think this would work. As duplicated template instances are also created in the same binding-level this would also error for template duplicates.

I compare the `loc` as templates will always have the same `loc` and user declared duplicated functions will have different `loc`s. I could probably use DECL_SOURCE_LOCATION instead, but maybe a better solution is following the dmd/linker behaviour more closely: Merge if the declaration is a template , otherwise error.